File: src/main/java/DroppableList.java
import javax.swing.*;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.awt.dnd.*;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.List;

public class DroppableList extends JList implements DropTargetListener, DragSourceListener, DragGestureListener {
    DropTarget dropTarget = new DropTarget(this, this);
    DragSource dragSource = DragSource.getDefaultDragSource();

    public DroppableList() {
        dragSource.createDefaultDragGestureRecognizer(this, DnDConstants.ACTION_COPY_OR_MOVE, this);
        setModel(new DefaultListModel());
    }

    public void dragDropEnd(DragSourceDropEvent DragSourceDropEvent) {
    }

    public void dragEnter(DragSourceDragEvent DragSourceDragEvent) {
    }

    public void dragExit(DragSourceEvent DragSourceEvent) {
    }

    public void dragOver(DragSourceDragEvent DragSourceDragEvent) {
    }

    public void dropActionChanged(DragSourceDragEvent DragSourceDragEvent) {
    }

    public void dragEnter(DropTargetDragEvent dropTargetDragEvent) {
        dropTargetDragEvent.acceptDrag(DnDConstants.ACTION_COPY_OR_MOVE);
    }

    public void dragExit(DropTargetEvent dropTargetEvent) {
    }

    public void dragOver(DropTargetDragEvent dropTargetDragEvent) {
    }

    public void dropActionChanged(DropTargetDragEvent dropTargetDragEvent) {
    }

    public synchronized void drop(DropTargetDropEvent dropTargetDropEvent) {
        try {
            Transferable tr = dropTargetDropEvent.getTransferable();
            if (tr.isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
                dropTargetDropEvent.acceptDrop(DnDConstants.ACTION_COPY_OR_MOVE);
                java.util.List fileList = (java.util.List) tr.getTransferData(DataFlavor.javaFileListFlavor);
                Iterator iterator = fileList.iterator();
                while (iterator.hasNext()) {
                    File file = (File) iterator.next();
                    ((DefaultListModel) getModel()).addAll(buildHashtable(file));
                }
                dropTargetDropEvent.getDropTargetContext().dropComplete(true);
            } else {
                System.err.println("Rejected");
                dropTargetDropEvent.rejectDrop();
            }
        } catch (IOException io) {
            io.printStackTrace();
            dropTargetDropEvent.rejectDrop();
        } catch (UnsupportedFlavorException ufe) {
            ufe.printStackTrace();
            dropTargetDropEvent.rejectDrop();
        }
    }

    public List<Hashtable> buildHashtable(File file) {
        List<Hashtable> tables = new ArrayList<>();
        if (file.isDirectory()) {
            File[] files = file.listFiles();
            if (files != null && files.length > 0) {
                for (File listedFile : files) {
                    tables.addAll(buildHashtable(listedFile));
                }
            }
        } else {
            Hashtable hashtable = new Hashtable();
            hashtable.put("name", file.getName());
            hashtable.put("path", file.getAbsolutePath());
            tables.add(hashtable);
        }
        return tables;
    }

    public void dragGestureRecognized(DragGestureEvent dragGestureEvent) {
        if (getSelectedIndex() == -1) return;
        Object obj = getSelectedValue();
        if (obj == null) {
            // Nothing selected, nothing to drag
            System.out.println("Nothing selected - beep");
            getToolkit().beep();
        } else {
            Hashtable table = (Hashtable) obj;
            FileSelection transferable = new FileSelection(new File((String) table.get("path")));
            dragGestureEvent.startDrag(DragSource.DefaultCopyDrop, transferable, this);
        }
    }
}

File: src/main/kotlin/DroppableList.kt
package de.griefed

import java.awt.datatransfer.DataFlavor
import java.awt.datatransfer.UnsupportedFlavorException
import java.awt.dnd.*
import java.io.File
import java.io.IOException
import java.util.*
import javax.swing.DefaultListModel
import javax.swing.JList

@Suppress("unused")
open class DroppableList : JList<Any?>(), DropTargetListener, DragSourceListener, DragGestureListener {

    var dragSource: DragSource = DragSource.getDefaultDragSource()

    init {
        dropTarget = DropTarget(this, this)
        dragSource.createDefaultDragGestureRecognizer(this, DnDConstants.ACTION_COPY_OR_MOVE, this)
        model = DefaultListModel<Any?>()
    }

    override fun dragDropEnd(DragSourceDropEvent: DragSourceDropEvent?) {
    }

    override fun dragEnter(DragSourceDragEvent: DragSourceDragEvent?) {
    }

    override fun dragExit(DragSourceEvent: DragSourceEvent?) {
    }

    override fun dragOver(DragSourceDragEvent: DragSourceDragEvent?) {
    }

    override fun dropActionChanged(DragSourceDragEvent: DragSourceDragEvent?) {
    }

    override fun dragEnter(dropTargetDragEvent: DropTargetDragEvent) {
        dropTargetDragEvent.acceptDrag(DnDConstants.ACTION_COPY_OR_MOVE)
    }

    override fun dragExit(dropTargetEvent: DropTargetEvent?) {
    }

    override fun dragOver(dropTargetDragEvent: DropTargetDragEvent?) {
    }

    override fun dropActionChanged(dropTargetDragEvent: DropTargetDragEvent?) {
    }

    @Synchronized
    override fun drop(dropTargetDropEvent: DropTargetDropEvent) {
        try {
            val tr = dropTargetDropEvent.transferable
            if (tr.isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
                dropTargetDropEvent.acceptDrop(DnDConstants.ACTION_COPY_OR_MOVE)
                val fileList = tr.getTransferData(DataFlavor.javaFileListFlavor) as MutableList<Any>
                val iterator: MutableIterator<Any> = fileList.iterator()
                while (iterator.hasNext()) {
                    val file = iterator.next() as File
                    (model as DefaultListModel<Any>).addAll(buildHashtable(file))
                }
                dropTargetDropEvent.dropTargetContext.dropComplete(true)
            } else {
                System.err.println("Rejected")
                dropTargetDropEvent.rejectDrop()
            }
        } catch (io: IOException) {
            io.printStackTrace()
            dropTargetDropEvent.rejectDrop()
        } catch (ufe: UnsupportedFlavorException) {
            ufe.printStackTrace()
            dropTargetDropEvent.rejectDrop()
        }
    }

    fun buildHashtable(file: File): MutableList<Hashtable<Any, Any>> {
        val tables: MutableList<Hashtable<Any, Any>> = ArrayList<Hashtable<Any, Any>>()
        if (file.isDirectory) {
            val files = file.listFiles()
            if (files != null && files.size > 0) {
                for (listedFile in files) {
                    tables.addAll(buildHashtable(listedFile) as Collection<Hashtable<Any, Any>>)
                }
            }
        } else {
            val hashtable: Hashtable<Any, Any> = Hashtable<Any, Any>()
            hashtable["name"] = file.name
            hashtable["path"] = file.absolutePath
            tables.add(hashtable)
        }
        return tables
    }

    override fun dragGestureRecognized(dragGestureEvent: DragGestureEvent) {
        if (selectedIndex == -1) return
        val obj = getSelectedValue()
        if (obj == null) {
            // Nothing selected, nothing to drag
            println("Nothing selected - beep")
            toolkit.beep()
        } else {
            val table = obj as Hashtable<Any, Any>
            val transferable = FileSelection(File(table["path"] as String))
            dragGestureEvent.startDrag(DragSource.DefaultCopyDrop, transferable, this)
        }
    }
}
